dafni/main_dafni.py

```diff
@@ -0,0 +1,210 @@
+                  variable["typestring"] == "Input"]
+
+        outputs = [Output(variable["name"], eval(variable["datatype"])) for variable in variables if
+                   variable["typestring"] == "Output"]
+
+        constraints = set()
+
+        for variable, _inputs in zip(variables, inputs):
+
+            if "constraint" in variable:
+                constraints.add(_inputs.z3 == variable["constraint"])
+
+    return inputs, outputs, constraints
+
+
+def main():
+    """
+    Main entrypoint of the script:
+    """
+    args = get_args()
+
+    # Step 0: Read in the runtime dataset(s)
+
+    try:
+
+        data_frame = pd.concat([pd.read_csv(d) for d in args.data_path])
+
+        # Step 1: Read in the JSON input/output variables and parse io arguments
+
+        variables_dict = read_variables(args.variables_path)
+
+        inputs, outputs, constraints = validate_variables(variables_dict)
+
+        # Step 2: Set up the modeling scenario and estimator
+
+        modelling_scenario = Scenario(variables=inputs + outputs, constraints=constraints)
+
+        modelling_scenario.setup_treatment_variables()
+
+        estimators = {"LinearRegressionEstimator": LinearRegressionEstimator}
+
+        # Step 3: Define the expected variables
+
+        expected_outcome_effects = {
+            "Positive": Positive(),
+            "Negative": Negative(),
+            "NoEffect": NoEffect(),
+            "SomeEffect": SomeEffect()}
+
+        # Step 4: Call the JSONUtility class to perform the causal tests
+
+        json_utility = JsonUtility(args.output_path, output_overwrite=True)
+
+        # Step 5: Set the path to the data.csv, dag.dot and causal_tests.json file
+        json_utility.set_paths(args.tests_path, args.dag_path, args.data_path)
+
+        # Step 6: Sets up all the necessary parts of the json_class needed to execute tests
+        json_utility.setup(scenario=modelling_scenario, data=data_frame)
+
+        # Step 7: Run the causal tests
+        test_outcomes = json_utility.run_json_tests(effects=expected_outcome_effects, mutates={}, estimators=estimators,
+                                                    f_flag=args.f)
+
+        # Step 8: Update, print and save the final outputs
+
+        for test in test_outcomes:
+
+            test.pop("estimator")
+
+            test["result"] = test["result"].to_dict(json=True)
+
+            test["result"].pop("treatment_value")
+
+            test["result"].pop("control_value")
+
+        with open(f"{args.output_path}", "w") as f:
+
+            print(json.dumps(test_outcomes, indent=2), file=f)
+
+        print(json.dumps(test_outcomes, indent=2))
+
+    except ValidationError as ve:
+
+        print(f"Cannot validate the specified input configurations: {ve}")
+
+    else:
+
+        print(f"Execution successful. Output file saved at {Path(args.output_path).parent.resolve()}")
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
For some old fashioned reason (joe heffer explained it once and I forgot), python files should end with a blank line haha